cmake/kokkos_functions.cmake

```diff
@@ -314,6 +314,151 @@ MACRO(kokkos_create_imported_tpl NAME)
   ENDIF()
 ENDMACRO()
 
+#
+# @MACRO: KOKKOS_FIND_HEADER
+#
+# Function that find a particular header. This searches custom paths
+# or default system paths depending on options. In constrast to CMake
+# default, custom paths are prioritized over system paths
+#
+# Usage::
+#
+#   KOKKOS_FIND_HEADER(
+#     <VAR_NAME>
+#     <HEADER>
+#     <TPL_NAME>
+#     ...
+#
+#   ``<VAR_NAME>``
+#
+#   The variable to define with the success or failure of the find
+#
+#   ``<HEADER>``
+#
+#   The name of the header to find
+#
+#   ``<TPL_NAME>``
+#
+#   The name of the TPL the header corresponds to
+#
+#   ``ALLOW_SYSTEM_PATH_FALLBACK``
+#
+#   If custom paths are given and the header is not found
+#   should we be allowed to search default system paths
+#   or error out if not found in given paths
+#
+#   ``PATHS``
+#
+#   Custom paths to search for the header
+#
+MACRO(kokkos_find_header VAR_NAME HEADER TPL_NAME)
+  CMAKE_PARSE_ARGUMENTS(TPL
+   "ALLOW_SYSTEM_PATH_FALLBACK"
+   ""
+   "PATHS"
+   ${ARGN})
+
+  SET(${HEADER}_FOUND FALSE)
+  SET(HAVE_CUSTOM_PATHS FALSE)
+  IF(NOT ${HEADER}_FOUND AND DEFINED ${TPL_NAME}_ROOT)
+    #ONLY look in the root directory
+    FIND_PATH(${VAR_NAME} ${HEADER} PATHS ${${TPL_NAME}_ROOT}/include NO_DEFAULT_PATH)
+    SET(HAVE_CUSTOM_PATHS TRUE)
+  ENDIF()
+
+  IF(NOT ${HEADER}_FOUND AND DEFINED KOKKOS_${TPL_NAME}_DIR)
+    #ONLY look in the root directory
+    FIND_PATH(${VAR_NAME} ${HEADER} PATHS ${KOKKOS_${TPL_NAME}_DIR}/include NO_DEFAULT_PATH)
+    SET(HAVE_CUSTOM_PATHS TRUE)
+  ENDIF()
+
+  IF (NOT ${HEADER}_FOUND AND TPL_PATHS)
+    #we got custom paths
+    #ONLY look in these paths and nowhere else
+    FIND_PATH(${VAR_NAME} ${HEADER} PATHS ${TPL_PATHS} NO_DEFAULT_PATH)
+    SET(HAVE_CUSTOM_PATHS TRUE)
+  ENDIF()
```

Comment:
Implicitly `TPL_PATHS` > `KOKKOS_${TPL_NAME}_DIR` >`${TPL_NAME}_ROOT` (where > indicates that a variable takes precedence over another)

Reply:
I set this as `ROOT > DIR > PATHS > SYSTEM`. I added some extra documentation for that. This seemed the most logical way of doing things?  This basically means user-provided > kokkos guesses > system defaults.  I can reorder, if anyone feels strongly this is not the right search order.